### PR TITLE
[Train] Torch data transfer automatic conversion

### PR DESCRIPTION
**Pull request by amogkam**
When saving a model in a checkpoint or reporting a model, the user has to manually extract out the module from DDP and move it to cpu so that it can be properly deserialized on the driver.

This PR adds functionality to automatically do the above so the user does not have to add this logic to their training script.

TODO: Possibly use this logic not just for checkpoints, but for return values as well.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/session.py [hunk 1]
```diff
@@ -195,6 +207,7 @@ def checkpoint(self, **kwargs):
             kwargs = {}
         else:
             kwargs = self._auto_fill_checkpoint_metrics(kwargs)
+            kwargs = self._encode_checkpoint_fn(kwargs)
 
         result = TrainingResult(TrainingResultType.CHECKPOINT, kwargs)
```
Comment (matthewdeng) [inline python/ray/train/session.py:212]: Is it intentional that this is only being done for checkpoints and not report? Maybe not as high priority because the users probably shouldn't be reporting the model, but I could imagine they might report some tensors?

File: python/ray/train/trainer.py [hunk 2]
```diff
@@ -566,14 +567,15 @@ class TrainingIterator:
...
-            self, backend_executor_actor: ActorHandle,
+            self, backend_executor_actor: ActorHandle, backend: Backend,
             train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
             run_dir: Path,
             dataset: Optional[Union[RayDataset, Dict[str, RayDataset]]],
             checkpoint_manager: CheckpointManager,
             checkpoint: Optional[Union[Dict, str, Path]],
             checkpoint_strategy: Optional[CheckpointStrategy]):
         self._backend_executor_actor = backend_executor_actor
+        self._backend = backend
```
Comment (matthewdeng) [inline python/ray/train/trainer.py:578]: Is this being used anywhere?

File: python/ray/train/backend.py [hunk 3]
```diff
@@ -21,6 +21,8 @@
 
 T = TypeVar("T")
 
+C = TypeVar("C")
```
Comment (matthewdeng) [inline python/ray/train/backend.py:24]: Maybe we can rename this into something more description, e.g:
```suggestion
EncodedCheckpointData = TypeVar("EncodedCheckpointData")
```

or

```suggestion
EncodedData = TypeVar("EncodedData")
```

File: python/ray/train/tests/test_gpu.py [hunk 4]
```diff
@@ -19,6 +20,40 @@ def ray_start_4_cpus_2_gpus():
...
+        # Move to GPU device.
+        model = ray.train.torch.prepare_model(model)
+
+        assert next(model.parameters()).is_cuda
+
+        ray.train.save_checkpoint(model=model)
+
+    trainer = Trainer("torch", num_workers=num_workers, use_gpu=True)
+    trainer.run(train_func)
+    trainer.shutdown()
```
Comment (matthewdeng) [inline python/ray/train/tests/test_gpu.py:40]: Do you need to check that `trainer.latest_checkpoint` is on CPU here?

File: python/ray/train/tests/test_trainer.py [hunk 5]
```diff
@@ -562,6 +563,25 @@ def train_func():
...
+        model = train.torch.prepare_model(model)
+
+        # Save DDP wrapped model.
+        train.save_checkpoint(model=model)
+
+    num_workers = 2
+    trainer = Trainer("torch", num_workers)
+    trainer.start()
+    trainer.run(train_fn)
+    trainer.shutdown()
```
Comment (matthewdeng) [inline python/ray/train/tests/test_trainer.py:582]: Same here, can you access the model from `trainer.latest_checkpoint` and assert that it's not wrapped in DDP?

File: python/ray/train/trainer.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/train/trainer.py:578]: Ah, use this to pass in `decode_checkpoint` to `_process_checkpoint`?

File: python/ray/train/torch.py [hunk 6]
```diff
@@ -135,3 +140,178 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+        for k, v in checkpoint.items():
+            if isinstance(v, DistributedDataParallel) and hasattr(v, "module"):
+                checkpoint[k] = v.module
+
+        # Convert the checkpoint dict to bytes, so that any GPU tensors that
+        # are in the checkpoint dict can be properly deserialized on the
+        # driver side, even if the driver does not have access to a GPU device.
+        _buffer = io.BytesIO()
+        torch.save(checkpoint, _buffer)
+        return _buffer.getvalue()
```
Comment (matthewdeng) [inline python/ray/train/torch.py:161]: Nice, love the comments here! In the future it would be good to include this information in the public docs as well so users aren't confused (by how magical it is).

File: python/ray/train/session.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/session.py]: Done!

File: python/ray/train/trainer.py (hunk 2 above)
Comment (amogkam) [reply python/ray/train/trainer.py:577]: Fixed!

File: python/ray/train/backend.py (hunk 3 above)
Comment (amogkam) [reply python/ray/train/backend.py]: Done!

File: python/ray/train/tests/test_trainer.py (hunk 5 above)
Comment (amogkam) [reply python/ray/train/tests/test_trainer.py]: If it passes all the way through, then it should be working fine otherwise it would have raised an error. But added a check just to be safe.

File: python/ray/train/tests/test_gpu.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/tests/test_gpu.py]: Done!

File: python/ray/train/backend.py [hunk 7]
```diff
@@ -66,13 +122,14 @@ class BackendExecutor:
     def __init__(
             self,
             backend_config: BackendConfig,
+            backend: Backend,
```
Comment (matthewdeng) [inline python/ray/train/backend.py:125]: nit: Undo this change since it doesn't make sense to have conflicting `BackendConfig` and `Backend`.

File: python/ray/train/trainer.py [hunk 8]
```diff
@@ -124,7 +121,8 @@ def __init__(
         self.logdir = self.create_logdir(logdir)
 
         # Setup executor.
-        backend_config = self._get_backend_config(backend)
+        self._backend_config = self._get_backend_config(backend)
+        self._backend = self._backend_config.backend_cls()
```
Comment (matthewdeng) [inline python/ray/train/trainer.py:125]: Let's rename the `Backend` class in a future PR to make this less confusing!

File: python/ray/train/tests/test_session.py [hunk 9]
```diff
@@ -165,17 +166,20 @@ def encode_checkpoint(checkpoint):
...
         world_size=1,
-        encode_checkpoint_fn=encode_checkpoint)
+        encode_data_fn=encode_checkpoint)
+
     session = get_session()
     session.start()
+    # Validate checkpoint is encoded.
+    validate_encoded()
+    # Validate report is encoded.
     validate_encoded()
```
Comment (matthewdeng) [inline python/ray/train/tests/test_session.py]: nit: To make this more explicit
```suggestion
    # Validate checkpoint is encoded.
    validate_encoded(TrainingResultType.CHECKPOINT)
    # Validate report is encoded.
    validate_encoded(TrainingResultType.REPORT)
```

and update `validate_encoded`:

```python
    def validate_encoded(type: TrainingResultType):
        next = session.get_next()
        assert next is not None
        assert next.type == type
        assert next.data["encoded"] is True
```

**Review by matthewdeng — APPROVED** [review]
LGTM!

File: python/ray/train/backend.py (hunk 7 above)
Comment (amogkam) [reply python/ray/train/backend.py]: Ok I made `Backend` into a singleton, so it's ok to instantiate any of the backends any number of times, and we don't need to pass around a single instance.

File: python/ray/train/tests/test_session.py (hunk 9 above)
Comment (amogkam) [reply python/ray/train/tests/test_session.py]: Done!

File: python/ray/train/trainer.py (hunk 8 above)
Comment (amogkam) [reply python/ray/train/trainer.py]: Yep makes sense!
